compiler/src/test/java/motif/compiler/TestHarness.java

```diff
@@ -77,92 +95,146 @@ private static boolean isTestDir(File file) {
+                : Collections.singletonList(externalClassesDir);
         Processor processor = new Processor();
-        Compilation compilation = compiler.compile(
-                externalOutputDir,
-                null,
-                processor,
-                testCaseDir,
-                noDagger);
+
+        KotlinCompilation.Result result = compile(
+                getFiles(testCaseDir),
+                classpaths,
+                options,
+                processor);
 
         if (isErrorTest) {
-            runErrorTest(compilation);
+            runErrorTest(result);
         } else {
-            runSuccessTest(compilation, processor.graph);
+            assertSucceeded(result);
+
+            ClassLoader classLoader;
+            if (externalClassesDir == null) {
+                classLoader = result.getClassLoader();
+            } else {
+                classLoader = new URLClassLoader(
+                        new URL[]{
+                                result.getOutputDirectory().toURI().toURL(),
+                                externalClassesDir.toURI().toURL()
+                        }, getClass().getClassLoader());
+            }
+            Class<?> testClass = classLoader.loadClass(testClassName);
+
+            if (testClass.getAnnotation(Ignore.class) == null) {
+                runSuccessTest(testClass, processor.graph);
+            }
+        }
+    }
+
+    private void assertSucceeded(KotlinCompilation.Result result) {
+        if (result.getExitCode() != KotlinCompilation.ExitCode.OK) {
+            assertWithMessage(result.getMessages()).fail();
+        }
+    }
+
+    private void assertFailed(KotlinCompilation.Result result) {
+        if (result.getExitCode() == KotlinCompilation.ExitCode.OK) {
+            assertWithMessage("Expected compilation to fail but encountered no errors.").fail();
         }
     }
 
-    private void runErrorTest(Compilation compilation) throws IOException {
-        assertThat(compilation).failed();
+    private KotlinCompilation.Result compile(
+            List<File> sourceFiles,
+            List<File> classpaths,
+            Map<String, String> aptArgs,
+            @Nullable Processor motifProcessor) {
+        KotlinCompilation compilation = new KotlinCompilation();
+        compilation.setAnnotationProcessors(getProcessors(motifProcessor));
+        compilation.setSources(sourceFiles.stream()
+                .map(SourceFile.Companion::fromPath)
+                .collect(Collectors.toList()));
+        compilation.setInheritClassPath(true);
+        compilation.setClasspaths(classpaths);
+        compilation.setKaptArgs(aptArgs);
+        compilation.setVerbose(false);
+
+        return compilation.compile();
+    }
+
+    private List<File> getFiles(File dir) throws IOException {
+        return getFiles(dir, file -> (file.getName().endsWith(".java") || file.getName().endsWith(".kt")) && !file.getName().equals("ScopeImpl.java"));
```

Comment:
kts?

Reply:
This is part of the test harness and we can assume all  kotlin files end with kt.